### ISSUE TITLE
US Export control information

### ISSUE THREAD

**Issue by zaddok** [rustls/rustls#1864]
TLDR: Please include export related documentation/statements in the project.

--

For people or organizations that are touched by the US system in some way. They must consider if they are deemed as "exporting" or "re-exporting" software with encryption functionality. Software that makes https requests is deemed exporting encryption software. In these cases notifications/declarations/reports and/or even permission must be requested from the US government.

I understand that usage of HTTPS functionality within the operating system itself is considered safe for export because it is considered "standard encryption". Only notification/reporting is needed when using standard encryption.

I am not a lawyer, but I understand that standard OS encryption or an _equivalent_ is also safe for export as long as notification/reporting is made. So it seems reasonable that rusttls is safe to be used by US impacted entities as long as the code of rusttls doesn't extend past standard OS functionality. But is this the case?

Requests:

1. It would be great if this project could update its readme to clarify if it is providing standard OS equivalent tls functionality (and thus safe to export) or if it also contains non standard functionality that is illegal to export without permission.  

2. If rusttls contains encryption functionality different or stronger to standard OS encryption, it would be good to have an ability to enable this using a feature flag, to allow users of rusttls to avoid the additional complexity of requesting export permission.

3. It is frustrating that we have to deal with US export rules, but it is what it is. The alternate default position of "who cares about draconian rules in other countries, stfu politicians" is reasonable. But, If this is the position of the rusttls project, it would be great if this could be clarified in the documentation. As, even a simple "stfu draconian countries" statement would help people evaluate if they can use the code or not.

**Comment by zaddok** [comment]
Depending on the contents of the rust-tis package, for US citizens or organizations, it may be illegal to (for example):

1. Upload load balancer binaries with rusttls to an overseas server.
2. Make https requests using rusttls in an iOS or android app. (Apple/Google report apps on behalf of even non US entities)
3. Carry a phone or laptop from inside to outside of the US if it has the rusttls code or compile binaries.

**Comment by cpu** [comment]
Hi there,

I don't think it's appropriate for the Rustls project to try and document the interplay between our software and US export regulations. Rustls itself provides no cryptography, instead relying on a cryptography provider (currently defaulting to [aws-lc-rs](https://github.com/aws/aws-lc-rs)). 

Most importantly, we're not lawyers and so our advice or documentation is likely to be incomplete, or inaccurate. We're also not in a position to keep up-to-date on changes that may affect the documentation. I would argue that how to determine what is or isn't "standard encryption functionality" by definitions acceptable to the US government is well outside of our wheelhouse.  I think this position is in harmony with our peers in the cryptography community (e.g. [Golang](https://github.com/golang/go/issues/66381)).

I believe users that are concerned about these regulations should consult their own lawyers in a context where they can consider the totality of the software they're deploying, its configuration, and its packaging/distribution.

---

Separately, speaking as an individual, I think the history of SSL/TLS as a technology has demonstrated that trying to impose legislation on the export of cryptography is **actively harmful** to security. The creation of "export grade" ciphers, and their affect on security outside of export situations (e.g. in [FREAK](https://freakattack.com/) and [LOGJAM](https://weakdh.org/) attacks) was a net negative on the security of all users, American or otherwise. If you're in a position where you're affected by this sort of requirement I would strongly encourage you to voice your dissent, hopefully pushing governing bodies to reevaluate.

**Comment by zaddok** [comment]
> I don't think it's appropriate for the Rustls project to try and document the interplay between our software and US export regulations. 

Understood, and this completely makes sense. In my opinion, what you just wrote is a basic template for something that should be in the project documentation somewhere visible. There is no way for an individual to know this is the state of the project otherwise. (Although, now that this ticket exists, this will help people looking for the information.)

**Comment by cpu** [comment]
> (Although, now that this ticket exists, this will help people looking for the information.)

I think this issue is discoverable enough to serve as a point of reference for our policy here.

Thanks for opening the issue. I'm going to close this for now since it seems like there isn't any more action to take and nobody else has volunteered other perspectives.
